### PR TITLE
Add a way to undo patchConsole

### PR DESCRIPTION
**Pull request by DeMoorJasper**
# ↪️ Pull Request

This PR takes a copy of the original console object before running patchConsole to allow us to undo patchConsole or directly address the original console for some edge cases where logging to the reporters is unwanted behaviour (for example logging errors from the reporters...)

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

File: packages/core/core/src/ReporterRunner.js [hunk 1]
```diff
@@ -60,9 +66,8 @@ export default class ReporterRunner {
           logger: new PluginLogger({origin: reporter.name}),
         });
       } catch (e) {
-        throw new ThrowableDiagnostic({
-          diagnostic: errorToDiagnostic(e, reporter.name),
-        });
+        // We shouldn't emit a report event here as we will cause infinite loops...
+        INTERNAL_ORIGINAL_CONSOLE.error(e);
```
Comment (DeMoorJasper) [inline packages/core/core/src/ReporterRunner.js]: Maybe we should first try to emit the error as a reporter event and then see if it returns and log it into the original console if it persists?

Not entirely sure if it's really necessary, however logging the error straight to the original console would remove any fancy error formatting we do... for reporters

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>packages/benchmarks/kitchen-sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 11.36s | +40.00ms |
| Cached | 9.73s | +4.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.ff215454.webp | 102.94kb | +0.00b | 228.00ms | +34.00ms ⚠️ |
| dist/modern/parcel.76ee4591.webp | 102.94kb | +0.00b | 227.00ms | +35.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.ff215454.webp | 102.94kb | +0.00b | 14.00ms | -1.00ms 🚀 |
| dist/modern/parcel.76ee4591.webp | 102.94kb | +0.00b | 12.00ms | -1.00ms 🚀 |
| dist/legacy/index.js | 243.00b | +0.00b | 19.00ms | -1.00ms 🚀 |
| dist/modern/index.js | 215.00b | +0.00b | 13.00ms | -2.00ms 🚀 |
| dist/legacy/index.css | 29.00b | +0.00b | 12.00ms | -1.00ms 🚀 |

</p></details>

<details><summary>packages/benchmarks/react-hn ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 16.26s | +835.00ms ⚠️ |
| Cached | 11.14s | +114.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/module/index.js | 34.02kb | +0.00b | 2.31s | +118.00ms ⚠️ |
| dist/main/index.js | 34.00kb | +0.00b | 2.31s | +117.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/module/index.js | 34.03kb | +0.00b | 11.00ms | +1.00ms ⚠️ |
| dist/main/index.js | 34.00kb | +0.00b | 16.00ms | +1.00ms ⚠️ |

</p></details>

<details><summary>packages/benchmarks/ak-editor 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/251541557827600905)**

File: packages/core/core/src/ReporterRunner.js [hunk 2]
```diff
@@ -46,6 +50,8 @@ export default class ReporterRunner {
 
     if (this.options.patchConsole) {
       patchConsole();
+    } else {
+      unpatchConsole();
```
Comment (wbinnssmith) [inline packages/core/core/src/ReporterRunner.js]: Why do we need to unpatch the console when patchConsole is false? How could it have been patched before this?

File: packages/core/logger/src/Logger.js [hunk 3]
```diff
@@ -176,6 +174,28 @@ export function patchConsole() {
   consolePatched = true;
 }
 
+export function unpatchConsole() {
```
Comment (wbinnssmith) [inline packages/core/logger/src/Logger.js]: can `patchConsole` return a Disposable [0] that does this instead?

[0] https://github.com/parcel-bundler/parcel/blob/b777024226c4f01efb30a2265dd45d90ae801a4d/packages/utils/events/src/Disposable.js

File: packages/core/logger/src/Logger.js [hunk 4]
```diff
@@ -140,15 +140,13 @@ export class PluginLogger {
 }
 
 let consolePatched = false;
+export const INTERNAL_ORIGINAL_CONSOLE = {...console};
```
Comment (wbinnssmith) [inline packages/core/logger/src/Logger.js]: Can this just be a direct reference to the original console?

File: packages/core/logger/src/Logger.js (hunk 3 above)
Comment (DeMoorJasper) [reply packages/core/logger/src/Logger.js]: I guess so but it wouldn't really change anything would it?

File: packages/core/core/src/ReporterRunner.js (hunk 2 above)
Comment (DeMoorJasper) [reply packages/core/core/src/ReporterRunner.js]: In tests it could have been patched in a previous test...

Probably could happen in any framework built on top of parcel or anything

File: packages/core/logger/src/Logger.js (hunk 4 above)
Comment (DeMoorJasper) [reply packages/core/logger/src/Logger.js]: we overwrite the console object values with our logger functions so I don't think that would actually do anything?

Comment (wbinnssmith) [reply packages/core/logger/src/Logger.js]: I didn't know we did this. Why override the individual functions?

Comment (DeMoorJasper) [reply packages/core/logger/src/Logger.js]: Probably because there's more to console as just the logger functions? Not sure, I think Devon wrote patchConsole
